src/cpu/o3/dyn_inst.hh

```diff
@@ -1066,6 +1066,23 @@ class DynInst : public ExecContext, public RefCounted
         const RegId& reg = si->destRegIdx(idx);
         assert(reg.is(MiscRegClass));
         setMiscReg(reg.index(), val);
+
+        /** If the MiscReg allows non-serializing updates, we
+         * can update the value immediately without waiting
+         * for commit, but only if the instruction is non
+         * speculative
+         */
+        if (!reg.regClass().isSerializing(reg)) {
+            assert(isNonSpeculative());
+
+            bool no_squash_from_TC = thread->noSquashFromTC;
+            thread->noSquashFromTC = true;
```

Comment:
Why is that needed?
Will the `setMisc` function do something weird?

Reply:
Copy pasted from the updateMiscRegs logic:

```
1092         // @todo: Pretty convoluted way to avoid squashing from happening when
1093         // using the TC during an instruction's execution (specifically for
1094         // instructions that have side-effects that use the TC).  Fix this.
1095         // See cpu/o3/dyn_inst_impl.hh.
```

IIRC there are cases where calling

thread->setMiscReg would cause a squash in the O3 model due to the following:

https://github.com/gem5/gem5/blob/stable/src/cpu/o3/thread_context.cc#L207C16-L207C26

The idea is that if you are changing anything through the TC interface, it means you are forcing a change in the execution context and therefore you need to squash the pipeline.